### PR TITLE
enhancements: support real-time kernels

### PR DESCRIPTION
**Pull request by miabbott**
This proposal covers adding support for real-time kernels to RHCOS and
the ability to select those kernels with the MCO.

### CODE REVIEW

**Comment by miabbott** [comment]
cc: @cgwalters @ashcrow @runcom @sinnykumari @imcleod @darkmuggle @mrguitar

File: enhancements/support-for-realtime-kernel.md [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+attempt to use the real-time kernel.  If users enable real-time kernel support
+without properly configured hardware, they can always remove the offending
+MachineConfig to return to the previous state.
+
+#### Lack of Full Support
+
+If there are additional changes required across the OCP product to properly support
+real-time kernel (i.e. container runtime, etc) which cannot be delivered at the
+same time, the real-time kernel packages can still be included as part of the
+`machine-os-content` image.  Additionally, exposing the tunable in the MachineConfig
```
Comment (darkmuggle) [inline enhancements/support-for-realtime-kernel.md]: Mind clarifying this? The part about lacking full support pivots to removing the tunable.

File: enhancements/support-for-realtime-kernel.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+- Defaulting to the real-time kernel or offering multiple boot images
+- Additional tuning of the RHCOS nodes after the real-time kernel is selected (handled by [Cluster Node Tuning Operator](https://github.com/openshift/cluster-node-tuning-operator))
+- Providing real-time kernel support to older versions (pre-4.4) of OCP
+- Supporting real-time kernel selection on non-RHCOS nodes
+
+## Proposal
+
+1.  Include the `kernel-rt` packages in the `machine-os-content` image
+2.  Provide tunable in MachineConfig that selects the type of kernel to use
+3.  Do initial tuning of RHCOS node after real-time kernel is applied
```
Comment (darkmuggle) [inline enhancements/support-for-realtime-kernel.md]: I think we need to clarify what "tuning" is affected. For example, FIPS and encryption?

File: enhancements/support-for-realtime-kernel.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+same time, the real-time kernel packages can still be included as part of the
+`machine-os-content` image.  Additionally, exposing the tunable in the MachineConfig
+spec can be removed.  Once support has been fully committed across the product,
+we can expose the tunable again.
+
+## Design Details
+
+### Providing the Packages
+
+The proposal is to include the `kernel-rt` packages in the `machine-os-content`
```
Comment (ashcrow) [inline enhancements/support-for-realtime-kernel.md]: nit: In the _standard_ `machine-os-content`. Basically denote it's not a special or different one.

File: enhancements/support-for-realtime-kernel.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+When the MCO parses a MachineConfig with `kernelType: realtime`, it shall instruct
+`rpm-ostree` on the RHCOS node to override the installed kernel with the `kernel-rt`
+package from the `machine-os-content` image.
+
+### Removing the Real-time Kernel
+
+When the MCO parses a MachineConfig with `kernelType: default`, it shall instruct `rpm-ostree`
+on the RHCOS node to remove any `kernel-rt` packages and use the default kernel.  If the
+`kernel-rt` packages are not present, it should be a no-op.
```
Comment (ashcrow) [inline enhancements/support-for-realtime-kernel.md]: Any logging/warning?

File: enhancements/support-for-realtime-kernel.md (hunk 1 above)
Comment (miabbott) [reply enhancements/support-for-realtime-kernel.md]: My logic was:

1) we prep RT kernel support in RHCOS + MCO
2) we discover there are other components that are not ready
3) mitigation:  keep the RT kernel in the `machine-os-content` and backout the MCO change, effectively hiding the knob from users

Does that help?

File: enhancements/support-for-realtime-kernel.md (hunk 2 above)
Comment (miabbott) [reply enhancements/support-for-realtime-kernel.md]: I believe this falls into the bucket of `tuned` profiles.  @cgwalters can you provide some details?

File: enhancements/support-for-realtime-kernel.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+the `machine-os-content` image.  This treats the `kernel-rt` package like any
+other RPM installed on the RHCOS nodes.
+
+### Test Plan
+
+- Test install of `kernel-rt` packages on single RHCOS node
+- Test removal of `kernel-rt` packages on single RHCOS node
+- Test `kernelType: realtime` on OCP cluster with RHCOS nodes using default kernel
+- Test `kernelType: default` on OCP cluster with RHCOS nodes using default kernel
+- Test `kernelType: default` on OCP cluster with RHCOS nodes already using real-time kernel
```
[the remark this answers is not in the document]
Comment (miabbott) [reply enhancements/support-for-realtime-kernel.md]: Got it; will update that test case.

File: enhancements/support-for-realtime-kernel.md (hunk 4 above)
Comment (miabbott) [reply enhancements/support-for-realtime-kernel.md]: I wasn't going to get into those details, but no harm in calling it out explicitly.

File: enhancements/support-for-realtime-kernel.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+- It should be noted that this increases the support matrix for the
+OpenShift support folks.  The amount of additional configurations that
+this could create (i.e. disk encryption on/off, different platforms,  etc.)
+is something that needs to be considered when reviewing the proposal.
+
+- Customers requiring use of custom kernel modules (or [SRO](https://github.com/zvonkok/special-resource-operator))
+for things like nVidia GPUs will need to rebuild the kernel modules
+to support the real-time kernel.
+
+- Customers requiring FIPS support **SHOULD NOT** use the real-time kernel.
```
Comment (ericavonb) [inline enhancements/support-for-realtime-kernel.md]: Can we handle this situation explicitly in the MCO? Like, the MCO can refuse to enable it (e.g. by setting the kerneltype=default on the rendered MC when fips=true)?

File: enhancements/support-for-realtime-kernel.md [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+When the OCP cluster is upgraded, the RHCOS nodes that are using the real-time
+kernel will also update the `kernel-rt` packages if updates are available in
+the `machine-os-content` image.  This treats the `kernel-rt` package like any
+other RPM installed on the RHCOS nodes.
+
+### Test Plan
+
+- Test install of `kernel-rt` packages on single RHCOS node
+- Test removal of `kernel-rt` packages on single RHCOS node
```
Comment (mike-nguyen) [inline enhancements/support-for-realtime-kernel.md]: Test upgrade of `kernel-rt` package on a single RHCOS node?  So we can possible catch breakage earlier than when it gets into OCP.  We'd need another version but same issue applies with updating OCP with a newer `kernel-rt`

File: enhancements/support-for-realtime-kernel.md [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+- Test `kernelType: realtime` on OCP cluster with RHCOS nodes using default kernel
+- Test `kernelType: default` on OCP cluster with RHCOS nodes using default kernel
+- Test `kernelType: default` on OCP cluster with RHCOS nodes already using real-time kernel
+- Test `kernelType: realtime` on OCP cluster with RHEL nodes
+- Test `kernelType: default` on OCP cluster with RHEL nodes
```
Comment (mike-nguyen) [inline enhancements/support-for-realtime-kernel.md]: Test Machine Config with no `kernelType`

File: enhancements/support-for-realtime-kernel.md (hunk 6 above)
Comment (miabbott) [reply enhancements/support-for-realtime-kernel.md]: Seems feasible.  @sinnykumari WDYT?

Comment (sinnykumari) [reply enhancements/support-for-realtime-kernel.md]: yes should be possible. MCO can print some message saying Not supported when FIPS is already enabled.

**Comment by miabbott** [comment]
Finally got around to making the updates :arrow_up:

**Comment by ashcrow** [comment]
Anyone object to merging?

**Comment by miabbott** [comment]
Can someone give a `/lgtm` here?

It's just a formality at this point; the `machine-os-content` already contains the RT kernel as described and the MCO has the functionality to [switch kernels](https://github.com/openshift/machine-config-operator/pull/1330)

**Pull request merged**